### PR TITLE
#5-Carousel

### PR DESCRIPTION
**Pull request by Mario-k4**
#5

### CODE REVIEW

File: components/Carousel/SlideCard.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
+'use client'
```
Comment (RevanchistX) [inline components/Carousel/SlideCard.tsx]: No need for use client here.

File: components/Carousel/SlideCard.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+type SlideCardProps = {
+    index: number;
+    name: string;
+};
+
+const SlideCard = ({ index, name }: SlideCardProps) => {
+    return (
+        <div
+            className={`rounded-3xl border-orange border-4 flex-0 flex-shrink-0 basis-[70%] max-w-[400px] lg:max-w-[500px] xl:max-w-[500px] min-w-0 ${index > 0 ? '-ml-6' : '-ml-4'} relative`}
```
Comment (RevanchistX) [inline components/Carousel/SlideCard.tsx]: Don't use custom values, just implement the ones that are available in tailwind.
You can use use this cheat sheet to navigate tailwind values easier.
https://nerdcave.com/tailwind-cheat-sheet

Comment (RevanchistX) [reply components/Carousel/SlideCard.tsx]: Additionally:
In scenarios like these `lg:max-w-[500px] xl:max-w-[500px]` , you have to remember that `lg` and `xl` are just shortcuts for `min-width:  xxx` which means that this declaration is redundant.

File: components/Carousel/EmblaCarouselArrowButton.tsx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            type="button"
+            {...restProps}
+        >
+            <svg className="embla__button__svg" viewBox="0 0 532 532">
+                <path
+                    fill="currentColor"
+                    d="M176.34 520.646c-13.793 13.805-36.208 13.805-50.001 0-13.785-13.804-13.785-36.238 0-50.034L330.78 266 126.34 61.391c-13.785-13.805-13.785-36.239 0-50.044 13.793-13.796 36.208-13.796 50.002 0 22.928 22.947 206.395 206.507 229.332 229.454a35.065 35.065 0 0 1 10.326 25.126c0 9.2-3.393 18.26-10.326 25.2-45.865 45.901-206.404 206.564-229.332 229.52Z"
+                />
+            </svg>
+            {children}
```
Comment (RevanchistX) [inline components/Carousel/EmblaCarouselArrowButton.tsx]: Buttons have no children.

File: components/Carousel/EmblaCarouselArrowButton.tsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    return {
+        prevBtnDisabled,
+        nextBtnDisabled,
+        onPrevButtonClick,
+        onNextButtonClick
+    }
+}
+
+type PropType = ComponentPropsWithRef<'button'>
```
Comment (RevanchistX) [inline components/Carousel/EmblaCarouselArrowButton.tsx]: Should have a more descriptive name.

File: components/Carousel/EmblaCarouselArrowButton.tsx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+export const PrevButton: React.FC<PropType> = (props) => {
```
Comment (RevanchistX) [inline components/Carousel/EmblaCarouselArrowButton.tsx]: You can refactor this so that both buttons can be done with the same logic if you extrapolate their common values

File: components/Carousel/SlideCard.tsx [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+            style={{ zIndex: 5 + index }}
+        >
+            <div className="relative rounded-3xl overflow-hidden sm:h-[600px] min-h-[300px]">
+                <Image
+                    src={`/images/slide-${index + 1}.jpg`}
+                    alt={`Slide ${index + 1}`}
+                    layout="fill"
```
Comment (RevanchistX) [inline components/Carousel/SlideCard.tsx]: `Layout` and `objectfit` are deprecated properties and throw console warnings.

Check out the documentation how to do it properly
https://nextjs.org/docs/pages/api-reference/components/image#fill

Check this stackoverflow discussion for how to approach the problem
https://stackoverflow.com/questions/64846858/how-to-use-tailwind-css-with-next-js-image

File: tsconfig.json [hunk 7]
```diff
@@ -32,9 +32,10 @@
     "next-env.d.ts",
     "**/*.ts",
     "**/*.tsx",
-    ".next/types/**/*.ts"
+    ".next/types/**/*.ts",
+    "declarations.d.ts"
```
Comment (RevanchistX) [inline tsconfig.json:36]: Why is this here?

**Review by RevanchistX — CHANGES_REQUESTED** [review]
Needs refactoring and some improvements

Additionally, images don't fill out the border and have weird gaps in them.
image

File: components/Calendar/Calendar.css
Comment (RevanchistX) [inline components/Calendar/Calendar.css:1]: Calendar functionality shouldn't be on this branch.

File: components/Calendar/ReactCalendar.tsx
Comment (RevanchistX) [inline components/Calendar/ReactCalendar.tsx:1]: Calendar functionality shouldn't be on this branch.

File: components/Carousel/SlideCard.tsx [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        <div
+            className={`rounded-3xl flex-0 flex-shrink-0 basis-[70%] max-w-sm lg:max-w-lg min-w-0 ${index > 0 ? '-ml-6' : '-ml-4'} relative`}
+            style={{ zIndex: 5 + index }}
+        >
+            <div className="relative border-orange border-4 rounded-3xl overflow-hidden sm:h-customHeight h-80">
+                <Image
+                    src={`/images/slide-${index + 1}.jpg`}
+                    alt={`Slide ${index + 1}`}
+                    fill={true}
+                    style={{ objectFit: "cover" }}
```
Comment (RevanchistX) [inline components/Carousel/SlideCard.tsx]: This can be a tailwind class

File: tailwind.config.ts [hunk 9]
```diff
@@ -27,6 +27,9 @@ const config: Config = {
             borderRadius: {
                 '4xl': '2rem',
             }
+        },
+        height: {
+            customHeight: '37.5rem'
```
Comment (RevanchistX) [inline tailwind.config.ts]: Needs more descriptive name.

File: components/constants/config.ts
Comment (RevanchistX) [inline components/constants/config.ts:1]: If this is related to the calendar, add it to the calendar folder.

File: components/Calendar/ReactCalendar.tsx
Comment (RevanchistX) [reply components/Calendar/ReactCalendar.tsx:1]: Additionally, you need to include the dependency in the `package.json` since this doesn't compile.

File: components/Carousel/EmblaCarouselArrowButton.tsx
Comment (RevanchistX) [inline components/Carousel/EmblaCarouselArrowButton.tsx:1]: You can simplify the name to `CarouselButtons.tsx`

File: components/Carousel/Carousel.tsx [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+                    {slideNames.map((name, index) => (
+                        <SlideCard key={index} name={name} index={index} />
+                    ))}
+                </div>
+            </div>
+        </section>
+    );
+};
+
+export default EmblaCarousel;
```
Comment (RevanchistX) [inline components/Carousel/Carousel.tsx]: Default export should match file name for consistency.

File: components/Carousel/Carousel.tsx [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+import SlideCard from './SlideCard';
+import { EmblaOptionsType } from 'embla-carousel';
+import {
+    PrevButton,
+    NextButton,
+    usePrevNextButtons
+} from './EmblaCarouselArrowButton';
+import useEmblaCarousel from 'embla-carousel-react';
+
+type PropType = {
```
Comment (RevanchistX) [inline components/Carousel/Carousel.tsx]: Needs more descriptive name.

File: components/Carousel/Carousel.tsx [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    slides: number[];
+    options?: EmblaOptionsType;
+};
+
+const EmblaCarousel = (props: PropType) => {
+    const slideNames = ['Арија', 'Реа', 'Фреја', 'Јоги', 'Дона'];
+    const { slides, options } = props;
```
Comment (RevanchistX) [inline components/Carousel/Carousel.tsx]: - `slides` variable / prop is not used anywhere. You're mapping them on `slideNames`, which are hardcoded in the component. That makes the entire declaration on `page.tsx` (Homepage) redundant. 
- `options` variable can be hardcoded into this component instead of being passed down from the parent.

This will be changed in the future to retrieve data from an API, but you can do a little bit of cleanup to prepare for that. Maybe it's best if you made some `json` dummy data which will contain the images info and just pass the `url` to this component which will fetch the url and render the component accordingly.

File: components/Carousel/Carousel.tsx [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+                        className="cursor-pointer w-8 h-8 sm:w-12 sm:h-12 z-10 text-black flex items-center justify-center"
+                        onClick={onNextButtonClick}
+                        disabled={nextBtnDisabled}
+                    />
+                </div>
+            </div>
+            <div className="overflow-hidden" ref={emblaRef}>
+                <div className="flex touch-pan-y">
+                    {slideNames.map((name, index) => (
+                        <SlideCard key={index} name={name} index={index} />
```
Comment (RevanchistX) [inline components/Carousel/Carousel.tsx]: `index` as `key` is not a good practice.

https://medium.com/@sriweb/avoid-using-array-index-as-a-key-in-react-components-1978799ed94b

File: components/Calendar/Calendar.css
Comment (RevanchistX) [reply components/Calendar/Calendar.css:1]: Additionally this css file needs to be imported into the app, otherwise it doesn't work.
Do this by adding `@import url(../components/Calendar/Calendar.css);` to `app/globals.css`

File: components/Carousel/Carousel.tsx (hunk 12 above)
Comment (RevanchistX) [reply components/Carousel/Carousel.tsx]: example json:
```
{
    "images": [
        {
            "id": "1",
            "name": "foo",
            "url": "../foo.jpg"
        },
        {
            "id": "2",
            "name": "bar",
            "url": "../bar.jpg"
        }
    ]
}
```

**Review by RevanchistX — CHANGES_REQUESTED** [review]
Remove Calendar logic from this branch. This will be reviewed in a separate PR.
For the Carousel logic, pay attention to the syntax and naming conventions. 
I noticed a lot of inconsistencies in the code, like missing `;` at end of lines.
Don't rush through making the changes as it makes it more prone to errors.

File: app/layout.tsx
Comment (RevanchistX) [inline app/layout.tsx:1]: Unnecessary changes.

File: package.json [hunk 14]
```diff
@@ -9,8 +9,11 @@
     "lint": "next lint"
   },
   "dependencies": {
+    "embla-carousel": "^8.1.6",
+    "embla-carousel-react": "^8.1.6",
     "next": "14.2.3",
     "react": "^18",
+    "react-calendar": "^5.0.0",
```
Comment (RevanchistX) [inline package.json]: Calendar doesn't belong in this PR

**Review by RevanchistX — CHANGES_REQUESTED** [review]
Don't forget to merge from staging.

Edit:
Images not showing on small screens.
image

**Comment by Mario-k4** [comment]
Merge conflicts have been resolved

**Review by RevanchistX — APPROVED** [review]
